src/shapes/Object/Object.ts

```diff
@@ -736,29 +734,29 @@ export class FabricObject<
       // i don't like this automatic initialization here
     } else if (key === 'shadow' && value && !(value instanceof Shadow)) {
       value = new Shadow(value);
-    } else if (key === 'dirty' && this.group && value) {
-      // a dirty child makes the parent dirty
-      // but a non dirty child will not make the parent non dirty.
-      // the parent could be dirty for some other reason
-      this.group.set('dirty', value);
     }
 
+    const isChanged = this[key as keyof this] !== value;
     this[key as keyof this] = value;
 
-    if (isChanged) {
-      const groupNeedsUpdate = this.group && this.group.isOnACache();
-      if (
-        (this.constructor as typeof FabricObject).cacheProperties.includes(key)
-      ) {
-        this.dirty = true;
-        groupNeedsUpdate && this.group!.set('dirty', true);
-      } else if (
-        groupNeedsUpdate &&
-        (this.constructor as typeof FabricObject).stateProperties.includes(key)
-      ) {
-        this.group!.set('dirty', true);
-      }
+    // invalidate caches
+    if (
+      isChanged &&
+      (this.constructor as typeof FabricObject).cacheProperties.includes(key)
+    ) {
+      this.dirty = true;
     }
+    // a dirty child makes the parent dirty.
+    // but a non dirty child does not make the parent not dirty.
+    // the parent could be dirty for some other reason.
+    this.parent &&
+      (this.dirty ||
+        (isChanged &&
+          (this.constructor as typeof FabricObject).stateProperties.includes(
+            key
+          ))) &&
+      this.parent._set('dirty', true);
```

Comment:
Here i have the same exact concern, group is our main property, while parent is just an exception sometimes.
So we need to check parent or group, with precedence to parent.

Also i think is ok to skip `const groupNeedsUpdate = this.group && this.group.isOnACache();` is a bit of hit and miss. Is true we don't care if the group is not cached to set it as dirty, since is just a boolean, but if the group is cached on a different cache than the object, at that point the invalidation isn't necessary and is a cost. ( reading the code today this check isn't made, but if say we are going always to flag a group as dirty we are cutting that away )

What is the reason behind changing the dirty flag? what brought you to change it?

Reply:
Yes you are right the separate cache is going to make the lower cache dirty anyway.
So the issue is just parent vs group